pkg/local_object_storage/engine/evacuate.go

```diff
@@ -0,0 +1,146 @@
+package engine
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/nspcc-dev/hrw"
+	meta "github.com/nspcc-dev/neofs-node/pkg/local_object_storage/metabase"
+	"github.com/nspcc-dev/neofs-node/pkg/local_object_storage/shard"
+	"github.com/nspcc-dev/neofs-node/pkg/util"
+	"go.uber.org/zap"
+)
+
+// EvacuateShardPrm represents parameters for the EvacuateShard operation.
+type EvacuateShardPrm struct {
+	shardID      *shard.ID
+	ignoreErrors bool
+}
+
+// EvacuateShardRes represents result of the EvacuateShard operation.
+type EvacuateShardRes struct {
+	count int
+}
+
+// WithShardID sets shard ID.
+func (p *EvacuateShardPrm) WithShardID(id *shard.ID) {
+	p.shardID = id
+}
+
+// WithIgnoreErrors sets flag to ignore errors.
+func (p *EvacuateShardPrm) WithIgnoreErrors(ignore bool) {
+	p.ignoreErrors = ignore
+}
+
+// Count returns amount of evacuated objects.
+func (p EvacuateShardRes) Count() int {
+	return p.count
+}
+
+const defaultEvacuateBatchSize = 100
+
+type pooledShard struct {
+	hashedShard
+	pool util.WorkerPool
+}
+
+var errMustHaveTwoShards = errors.New("amount of shards must be > 2")
+
+// Evacuate moves data from one shard to the others.
+// The shard being moved must be in read-only mode.
+func (e *StorageEngine) Evacuate(prm EvacuateShardPrm) (EvacuateShardRes, error) {
+	sid := prm.shardID.String()
+
+	e.mtx.RLock()
+	sh, ok := e.shards[sid]
+	if !ok {
+		e.mtx.RUnlock()
+		return EvacuateShardRes{}, errShardNotFound
+	}
+
+	if len(e.shards) < 2 {
+		e.mtx.RUnlock()
+		return EvacuateShardRes{}, errMustHaveTwoShards
+	}
+
+	// We must have all shards, to have correct information about their
+	// indexes in a sorted slice and set appropriate marks in the metabase.
+	// Evacuated shard is skipped during put.
+	shards := make([]pooledShard, 0, len(e.shards))
+	for id := range e.shards {
+		shards = append(shards, pooledShard{
+			hashedShard: hashedShard(e.shards[id]),
+			pool:        e.shardPools[id],
+		})
+	}
+	e.mtx.RUnlock()
+
+	weights := make([]float64, 0, len(shards))
+	for i := range shards {
+		weights = append(weights, e.shardWeight(shards[i].Shard))
+	}
+
+	if !sh.GetMode().ReadOnly() {
+		return EvacuateShardRes{}, shard.ErrMustBeReadOnly
+	}
+
+	var listPrm shard.ListWithCursorPrm
+	listPrm.WithCount(defaultEvacuateBatchSize)
+
+	var c *meta.Cursor
+	var res EvacuateShardRes
+	for {
+		listPrm.WithCursor(c)
+
+		// TODO (@fyrchik): #1731 this approach doesn't work in degraded modes
+		//  because ListWithCursor works only with the metabase.
+		listRes, err := sh.Shard.ListWithCursor(listPrm)
+		if err != nil {
+			return res, nil
```

Comment:
is that (listing error) the only success return from the infinite loop?

Reply:
Yes, fixed error check